tensorflow/compiler/mlir/tensor/transforms/legalize_utils.cc

```diff
@@ -0,0 +1,249 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <algorithm>
+#include <iostream>
+#include <numeric>
+
+#include "mlir/Dialect/Arith/IR/Arith.h"  // from @llvm-project
+#include "mlir/Dialect/Linalg/IR/Linalg.h"  // from @llvm-project
+#include "mlir/Dialect/Math/IR/Math.h"  // from @llvm-project
+#include "mlir/Dialect/SCF/IR/SCF.h"  // from @llvm-project
+#include "mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/tensor/transforms/legalize_utils.h"
+
+namespace mlir {
+namespace tensor {
+
+namespace {
+
+// Return the total size of the given tensor if it is ranked and all of its
+// dimensions have static size. Otherwise return nullopt. The given value is
+// expected to be a tensor.
+std::optional<int64_t> getTensorStaticSize(Value tensor) {
+  auto tensor_type = tensor.getType().cast<TensorType>();
+  auto ranked_tensor_type = tensor_type.dyn_cast<RankedTensorType>();
+  if (!ranked_tensor_type)
+    return std::nullopt;
+
+  int64_t size = 1;
+  for (auto dim_size : ranked_tensor_type.getShape()) {
+    if (dim_size == ShapedType::kDynamic)
+      return std::nullopt;
+    size *= dim_size;
+  }
+  return size;
+}
+
+// If the given SSA value is a constant tensor of integers, return the
+// corresponding values. Otherwise, nullopt.
+std::optional<SmallVector<int64_t>> getConstantShape(Value shape) {
+  DenseIntElementsAttr shape_attr;
+  if (!matchPattern(shape, m_Constant(&shape_attr)))
+    return std::nullopt;
+
+  return llvm::map_to_vector(shape_attr.getValues<IntegerAttr>(), [](IntegerAttr attr) {
+    return attr.getInt();
+  });
+}
+
+// Return whether the given shape has a wildcard (element set to -1).
+bool constantShapeHasWildcard(ArrayRef<int64_t> shape) {
+  return llvm::any_of(shape, [](auto size) {
+    return size == -1;
+  });
+}
+
+// Substitute the occurrence of -1 in the given shape with a new positive value
+// such that the product of all values in 'input_size' is equal to the product
+// in all values of 'shape'.
+SmallVector<int64_t> substituteConstantShapeWildcard(
+    ArrayRef<int64_t> input_size, ArrayRef<int64_t> shape) {
+  auto input_size_product = std::accumulate(input_size.begin(), input_size.end(), 1, std::multiplies<int64_t>());
```

Comment:
This is only possible if only one unknown correct? (I think this exact function is inside tflite converter too)

Reply:
Correct, only one occurrence of -1 is allowed. It's not explicitly stated in the spec, but this implementation assumes UB if more than one occurrence is present. I've been digging around and couldn't find an existing appropriate alternative. Do you have anything specific in mind?